### ISSUE TITLE
Type-side `instanceof` keyword and functionality

### ISSUE THREAD

**Issue by craigphicks** [microsoft/TypeScript#58181]
### 🔍 Search Terms

type side `instanceof` keyword 
mapping constructors

related issues:
#31311 #42534
#202 (seems indirectly related, but has been linked from #31311 and #42534)

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

A new type-side keyword `instanceof` is proposed to allow an interface to be declared as mapping to a specific class or specific constructor variable.  When type-checking for assignability of said interface, the type checker would not rely on structural duck-typing, but instead rely on the tracked class/variable from the object was constructed.

It directly addresses runtime errors resulting from runtime use of run-side `instanceof` that are not caught by the type checker.  However, it may also be useful for other reasons/purposes, as shown in "Motivating Examples and discussed in "Use Cases".

Of course absence of this feature is not a bug, as the overwhelming majority of TypeScript use cases *require not having having this feature enabled*.  However, there are a minority of use cases where this feature would be essential and/or useful, and the absence of this feature is a limitation.

### Example 1.1:

Library side using type-side `instanceof`:
```ts
class FooPrivatish {
    x: number;
    constructor(x:number){
        this.x = x;
    }
    valueOf(){
        return this.x;
    }
}

// Notice `instanceof FooPrivatish` V V V V V V V V V V V
export interface FooInstanceofable instanceof FooPrivatish {
    valueOf(): number;
}
export interface FooInstanceofableConstructor {
    new(x: number): FooInstanceofable;
}
export var FooInstanceofable: FooInstanceofableConstructor = FooPrivatish;
export function processFooInstanceofable(foo: FooInstanceofable): number {
    if (!(foo instanceof FooPrivatish)) throw new Error('foo is not instanceof class FooPrivatish');
    return foo.valueOf()
}
```

Library side with existing behavior, doesn't use type-side `instanceof`:
```ts
class FooLikePrivatish {
    x: number;
    constructor(x:number){
        this.x = x+1;
    }
    valueOf(){
        return this.x;
    }
}
interface FooLike {
    valueOf(): number;
}
export interface FooLikeConstructor {
    new(x: number): FooLike;
}
export const FooLike: FooLikeConstructor = FooLikePrivatish;

function processFooLike(foo: FooLike): number {
    return foo.valueOf()
}
```

User side, shows difference between existing and proposed behavior, and how they coexist:
```ts
import {....} from '...';
import {....} from '...';

const fooInstanceofable = new FooInstanceofable(0);
const fooLike = new FooLike(0);

processFooInstanceofable(fooInstanceofable); // no ts error
processFooInstanceofable(fooLike); // ts error - fooLike is not an instanceof FooInstanceofable

processFooLike(fooInstanceofable); // no ts error (existing behavior)
processFooLike(fooLike); // no ts error

fooLike satisifies FooInstanceofable; // ts error
fooInstanceofable satisfies fooLike ; // no ts error (existing behavior)

const foo = Math.random() > 0.5 ? fooInstanceofable : fooLike;
foo satisifies FooInstanceofable; // ts error
if (foo instanceof FooInstanceofable) {
    foo satisifies FooInstanceofable; // no ts error
}
```

### Example 1.2:

Unlike Example 1.1, this example shows `intanceof` being used to map to a constructor variable rather than an actual class.

Library side:
```ts
interface Boolean {
    /** Returns the primitive value of the specified object. */
    valueOf(): boolean;
}

interface BooleanConstructor {
    new (value?: any): Boolean;
    <T>(value?: T): boolean;
    readonly prototype: Boolean;
}

declare const Boolean: BooleanConstructor;  // No corresponding exported class.

interface BooleanInstanceofable instanceof Boolean {
    valueOf(): boolean;
}

```

User or other library side:
```ts
interface Array {
  //...
  filter<T>(x:T, f: BooleanInstanceofable, ...): Exclude<T, 0|-0|typeof NaN|null|undefined|false|"">[];
  filter<T>(x:T, f: (x:T, f: ...)=>any, ...): T[];
  //...
}
```
Note that if the argument type of `f` were just `Boolean`, then anything that has a `valueOf` method would pass the type check, even if it were a different class with different behavior, and the narrowed output type would therefore be wrong.

### 💻 Use Cases

1.  #42534 - in which a run time error is not caught by the type checker. That is working as expected but this proposal would allow a way for the user to make the type checker catch the error. (Also see [comment](https://github.com/microsoft/TypeScript/issues/202#issuecomment-2049461910)).  Similar to #31311.

2. The [workaround](https://github.com/microsoft/TypeScript/issues/31311#issuecomment-490690695) for #31311 requires a normal structural type check has unknown complexity, but type-side `instanceof` is known low fixed time complexity.

3. For an empty interface (representing a hidden class) which currently has no type checking at all, this proposal would allow the user to add type checking.

4. Example 1.2 above:  That can't be done presently because anything Boolean-like would pass the assignment test but it could be a different class with different behavior.

**Comment by saltman424** [comment]
> This auxilliary proposal allows instanceof to be placed prior to class or constructor variable, acting a qualifier.
>
> (Note: I now think this proposal might be better than the main proposal in terms of user ease and simplicity - while solving the same problems).

Agreed. I don't think restricting `instanceof` to only be useable with interfaces would be desirable or necessary. However, there are some interesting cases, particularly when considering using `extends`:

For example, using this setup:
```ts
class A { ... }
class B extends A { ... }
class C { ... }

// Basic examples of `instanceof ...` being used where you would use other types
const a1: instanceof A = new A()
const a2: typeof a1 = a1
type InstanceOfA = instanceof A
interface AlsoInstanceOfA extends InstanceOfA {
}
const a3: AlsoInstanceOfA = new A()
```

These accurately represent the runtime, so they probably should be allowed:
```ts
const b1: instanceof A = new B()
const b2: (instanceof A) & B = new B()
const b3: (typeof a1) & B = new B()
const b4: (instanceof A) & (instanceof B) = new B()
type AB = (instanceof A) & (instanceof B) // could be simplified under-the-hood to `instanceof B` because B extends A
```

And these are not possible, so they should be `never`:
```ts
type BC = (instanceof B) & (instanceof C)
type IndirectAC = AlsoInstanceOfA & (instanceof C)
```

Also, presumably mapped types reduce an `instanceof` type to a plain structural type. E.g.
```ts
type MappedA = { [K in keyof InstanceOfA]: InstanceOfA[K] }
const a1: MappedA = new A()
const a2: InstanceOfA = a1 // Error: MappedA matches the structure of InstanceOfA but may not be an instance of A
```

#### Side Note: Boolean
@craigphicks  I don't think your examples 1.2 and 2.3 do what you intend. My understanding is that would match `filter(new Boolean())` (since `new Boolean() instanceof Boolean === true`) when I believe you are intending to match `filter(Boolean)` (but `Boolean instanceof Boolean === false`).

**Comment by saltman424** [comment]
`typeof instanceof A` is not valid since `instanceof A` is a type, and `typeof` can only be used on a value. Assuming you meant something like:
```ts
const a: instanceof A
type X = typeof a
```
Then `X` would be `instanceof A` as `a` has a type of `instanceof A`

**Comment by saltman424** [comment]
@craigphicks I think I understand your point: TypeScript currently only has structural types, so right now, `typeof a` can equally be interpreted as "the *structural* type of `a`" and "the *entire* type of `a`". So if this proposal is enacted, there has to be a decision on which of those two interpretations is correct. With that being said, my response is basically:
1. I think it would break a lot of things if `typeof` only retrieved the structural type information (see another example below)
2. I can't think of anything that wouldn't work as expected if `typeof` included all of the type information
3. I don't think it is confusing to developers if the official position was: `typeof x` is just the type of `x`, whatever that type may be (structural or non-structural)

Another edge case if `typeof` didn't support non-structural types:
```ts
function f<T>(foo: T): void {
  let temp: typeof foo = foo
  // Today this works. Would this stop working because `T` might be an `instanceof` type, like in the call below?
  foo = temp
}

f<instanceof A>(new A())
```

**Comment by petermakeswebsites** [comment]
I would really appreciate this feature.

Using JS's `instanceof` is part of my workflow quite often. But when `instanceof's` are inside of functions, it's difficult or impossible to make an appropriate return type. Here's a very rudimentary version of the issue:

```typescript
class Foo {
  baz: string
}

class Bar {
  baz: string
}

const something : Foo | Bar = new Bar()

// TypeScript clearly has the ability built-in to differentiate Foo from Bar under the hood when using instanceof
if (something instanceof Bar) {
  something // TypeScript correctly narrows to Bar
} else {
  something // Typescript correctly narrows to Foo
}

const errorIfBar<T>(obj : T) : ??? { // here something along the lines of `T excluding instanceof Bar` would suffice perfectly
  if (obj instanceof Bar) throw new Error("got a Bar!")
  return obj
}
```

Whatever TypeScript is doing under the hood from the `if ( ... instanceof ... ) { .. }` in terms of type narrowing, it would be very handy to be able to use that.

In the above example, when you want to explicitly specify a return type of a function that uses `instanceof`, structural typing doesn't work because, for example, `Exclude<T, Bar>` would exclude both because both `Foo` and `Bar` are the same shape.

**Comment by musjj** [comment]
Here's another use case I bumped into. I wanted to make an immutable map between class constructors and instances, with type-safe accessors:

```ts
class Foo {}
class Bar {}
class Baz {}

interface AnyClassInstance {
  constructor: Function;
};

class Collection<T extends AnyClassInstance[]> {
    private collection: Map<Function, any>;

    constructor(collection: T) {
        this.collection = new Map();
        for (const c of collection) {
            this.collection.set(c.constructor, c);
        }
    }

    get<U extends T[number]>(t: new (...args: any) => U) {
        return this.collection.get(t) as U;
    }
}

const collection: Collection<[Foo, Bar]> = new Collection([new Foo(), new Bar()]);

const foo = collection.get(Foo);
const bar = collection.get(Bar);
const baz = collection.get(Baz); // this should not be allowed, but it passes the typechecker!
```

Because of how aggressively structurally-typed TypeScript is, `Baz` is considered a subset of `[Foo, Bar]` here. There's no way to enforce that the `U` must be one of the class instances in `T`.
